library/std/src/sys/uefi/common.rs

```diff
@@ -0,0 +1,488 @@
+        };
+
+        if r.is_error() { Err(status_to_io_error(r)) } else { Ok(()) }
+    }
+
+    #[inline]
+    pub(crate) fn as_raw_event(&self) -> r_efi::efi::Event {
+        self.inner.as_ptr()
+    }
+}
+
+impl Drop for Event {
+    fn drop(&mut self) {
+        if let Some(boot_services) = uefi::env::get_boot_services() {
+            // Always returns EFI_SUCCESS
+            let _ = unsafe { ((*boot_services.as_ptr()).close_event)(self.inner.as_ptr()) };
+        }
+    }
+}
+
+// A type to make working with UEFI DSTs easier
+// The Layout of this type has to be explicitly supplied
+// Inspiered by Box
+pub struct VariableBox<T> {
+    inner: Unique<T>,
+    layout: Layout,
+}
+
+impl<T> VariableBox<T> {
+    #[inline]
+    pub const unsafe fn from_raw(inner: *mut T, layout: Layout) -> VariableBox<T> {
+        VariableBox { inner: unsafe { Unique::new_unchecked(inner) }, layout }
+    }
+
+    #[inline]
+    pub const fn into_raw_with_layout(b: VariableBox<T>) -> (*mut T, Layout) {
+        let (leaked, layout) = VariableBox::into_unique(b);
+        (leaked.as_ptr(), layout)
+    }
+
+    pub fn try_new_uninit(layout: Layout) -> Result<VariableBox<MaybeUninit<T>>, AllocError> {
+        let inner = Global.allocate(layout)?.cast();
+        unsafe { Ok(VariableBox::from_raw(inner.as_ptr(), layout)) }
+    }
+
+    #[inline]
+    pub const fn leak<'a>(b: Self) -> &'a mut T {
+        unsafe { &mut *mem::ManuallyDrop::new(b).inner.as_ptr() }
+    }
+
+    #[inline]
+    pub const fn into_unique(b: Self) -> (Unique<T>, Layout) {
+        let layout = b.layout;
+        (Unique::from(VariableBox::leak(b)), layout)
+    }
+
+    #[inline]
+    pub fn layout(&self) -> Layout {
+        self.layout
+    }
+}
+
+impl<T> VariableBox<MaybeUninit<T>> {
+    #[inline]
+    pub unsafe fn assume_init(self) -> VariableBox<T> {
+        let (raw, layout) = VariableBox::into_raw_with_layout(self);
+        unsafe { VariableBox::from_raw(raw as *mut T, layout) }
+    }
+}
+
+impl<T> AsRef<T> for VariableBox<T> {
+    #[inline]
+    fn as_ref(&self) -> &T {
+        unsafe { self.inner.as_ref() }
+    }
+}
+
+impl<T> AsMut<T> for VariableBox<T> {
+    #[inline]
+    fn as_mut(&mut self) -> &mut T {
+        unsafe { self.inner.as_mut() }
+    }
+}
+
+impl<T> Deref for VariableBox<T> {
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl<T> DerefMut for VariableBox<T> {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.as_mut()
+    }
+}
+
+impl<T> Drop for VariableBox<T> {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe {
+            Global.deallocate(NonNull::new_unchecked(self.inner.as_ptr()).cast(), self.layout)
+        }
+    }
+}
+
+// Locate handles with a particular protocol GUID
+/// Implemented using `EFI_BOOT_SERVICES.LocateHandles()`
+pub(crate) fn locate_handles(mut guid: Guid) -> io::Result<Vec<NonNull<crate::ffi::c_void>>> {
```

Comment:
Why doesn't this just return `Handle` instead of `NonNull<..>`? AFAIK, a null pointer might be a perfectly valid handle, depending on the implementation.

Reply:
Oh, I didn't know that. Methods like `InstallProtocol` are supplied a `NULL` handle when a new handle needs to be allocated. So I assumed that an allocated handle would be non-null.

Reply:
Ah yes, I see. Never mind then!